examples/static_map/custom_type.cu

```diff
@@ -0,0 +1,128 @@
+
+#include <cuco/static_map.cuh>
+
+// User-defined key type
+#ifdef BUILD_PASCAL_CODE
+// Manual alignment required due to WAR libcu++ bug where cuda::atomic fails for underaligned types
+struct alignas(8) custom_key_type {
+  int32_t a;
+  int32_t b;
+
+  __host__ __device__ custom_key_type() {}
+  __host__ __device__ custom_key_type(int32_t x) : a{x}, b{x} {}
+
+  // Device equality operator is mandatory
+  __device__ bool operator==(custom_key_type const& other) const
+  {
+    return a == other.a and b == other.b;
+  }
+};
+#else
+// Key type larger than 8B only supported for sm_70 and up
+struct custom_key_type {
+  int32_t a;
+  int32_t b;
+  int32_t c;
+
+  __host__ __device__ custom_key_type() {}
+  __host__ __device__ custom_key_type(int32_t x) : a{x}, b{x}, c{x} {}
+
+  // Device equality operator is mandatory
+  __device__ bool operator==(custom_key_type const& other) const
+  {
+    return a == other.a and b == other.b and c == other.c;
+  }
+};
+#endif
+
+// User-defined value type
+// Manual alignment required due to WAR libcu++ bug where cuda::atomic fails for underaligned types
+struct alignas(8) custom_value_type {
+  int32_t f;
+  int32_t s;
+
+  __host__ __device__ custom_value_type() {}
+  __host__ __device__ custom_value_type(int32_t x) : f{x}, s{x} {}
+};
+
+// User-defined device hash callable
+struct custom_hash {
+  template <typename key_type>
+  __device__ uint32_t operator()(key_type k)
+  {
+    return k.a;
+  };
+};
+
+// User-defined device key equal callable
+struct custom_key_equals {
+  template <typename key_type>
+  __device__ bool operator()(key_type const& lhs, key_type const& rhs)
+  {
+    return lhs.a == rhs.a;
+  }
+};
+
+int main(void)
+{
+  constexpr std::size_t num_pairs = 80'000;
+
+  // Set emtpy sentinels
+  auto const empty_key_sentinel   = custom_key_type{-1};
+  auto const empty_value_sentinel = custom_value_type{-1};
+
+  thrust::device_vector<thrust::pair<custom_key_type, custom_value_type>> pairs(num_pairs);
+  // Create a sequence of 80'000 pairs
+  thrust::transform(
+    thrust::make_counting_iterator<int>(0),
+    thrust::make_counting_iterator<int>(num_pairs),
+    pairs.begin(),
+    [] __device__(auto i) { return thrust::make_pair(custom_key_type{i}, custom_value_type{i}); });
+
+  // Construct a map with 100,000 slots using the given empty key/value sentinels. Note the
+  // capacity is chosen knowing we will insert 80,000 keys, for an load factor of 80%.
+  cuco::static_map<custom_key_type, custom_value_type> map{
+    100'000, empty_key_sentinel, empty_value_sentinel};
+
+  // Inserts all pairs into the map by using the custom hasher and custom equality callable
+  map.insert(pairs.begin(), pairs.end(), custom_hash{}, custom_key_equals{});
+
+  // Reproduce inserted keys
+  auto insert_keys =
+    thrust::make_transform_iterator(thrust::make_counting_iterator<int32_t>(0),
+                                    [] __device__(auto i) { return custom_key_type{i}; });
```

Comment:
Instead of materializing the inputs in device memory, just use iterators like this for the original insert too.